raft-tech/TANF-app
# pipeline filtering

## When merging, update the build-and-test required check to build-and-test-all

## Summary of Changes
Leverages GitHub Actions to trigger more nuanced workflows in CircleCI based upon branches and paths. Only the build and test workflows for the part of code that was changed should run.
Pull requests and changes to the pipeline should still run the full build and test workflows.
CircleCI still builds main, master, and release branches on HHS side. But doesn't trigger for dev team branches.
Pull request closes #2457 

## How to Test
Change code and push, or make a PR. Only workflows that pertain to the changed materials should get run.
- only backend tests should be run if code is changed in tdrs-backend/ files
- only frontend tests should be run if code is changed in tdrs-frontend/ files
- infrastructure pipelines should only get run when making changes to files that pertain to infrastructure
- all tests should run if scripts that builds and pipelines use get changed, or changes to the pipelines themselves
- all tests should run when a PR is ready for review 

Since this is a pipeline change, testing is difficult to test besides against the actual pipeline.
[GitHub actions ](https://github.com/raft-tech/TANF-app/actions) has been tested so that code pushed to unwanted branches does not run (main, master, release/**), but is picked up by [CircleCI](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13731)
And likewise, CircleCI shouldn't run against [non-tracked branches](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13738) but instead be kicked off by Actions for [frontend/backend specific code changes](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13740) as well as [build-all for PRs](https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13739)

**on the HHS side, pipelines should not be affected**
- release/*, main, and master pushes should still build the CircleCI workflow (bypassing Actions)
tested on [forked repository](https://github.com/George-Hudson/TANF-app/branches)
tested on separate [CircleCI project](https://app.circleci.com/pipelines/github/George-Hudson/TANF-app)

## Deliverables
Should be able to see only workflows being run in circleci that pertain to the code change

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] [**code change to tdrs/backend only triggers backend tests**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874720916
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13368
+ [x] [**code change to tdrs/frontend only triggers frontend tests**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874770990
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13370
+ [x] [**code change to scropts or pipeline changes triggers all tests**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874701766
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13366
+ [x] [**infrastructure deploys are only triggered when code managing the infrastructure changes**]
   https://github.com/raft-tech/TANF-app/actions/runs/4874876141
   https://app.circleci.com/pipelines/github/raft-tech/TANF-app/13372

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

No changes to tested code were made

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)


### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)



### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)


### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)


### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)
